File: src/ui/mcp_client.js
/* MCP Streamable HTTP client — transport layer. No DOM access. No UI state. */

/**
 * @private
 * @param {string} block
 * @returns {string|null}
 */
export function extractDataField(block) {
  for (const line of block.split('\n')) {
    if (line.startsWith('data:')) {
      let val = line.slice(5);
      if (val.startsWith(' ')) val = val.slice(1);
      return val;
    }
  }
  return null;
}

/**
 * @private
 * @param {string} buf
 * @returns {{ event: object|null, remainder: string }}
 */
export function extractFirstDataEvent(buf) {
  let pos = 0;

  while (true) {
    const doubleNl = buf.indexOf('\n\n', pos);
    if (doubleNl === -1) {
      return { event: null, remainder: buf.slice(pos) };
    }

    const block = buf.slice(pos, doubleNl);
    pos = doubleNl + 2;

    if (!block.trim() || block.startsWith(':')) continue;

    const data = extractDataField(block);
    if (data) {
      return { event: JSON.parse(data), remainder: buf.slice(pos) };
    }
  }
}

export class McpClient {
  constructor() {
    this.sessionId = null;
    this.protocolVersion = null;
    this.requestId = 0;
    this.connected = false;
  }

  async initialize() {
    const body = {
      jsonrpc: '2.0',
      id: ++this.requestId,
      method: 'initialize',
      params: {
        protocolVersion: '2025-11-25',
        capabilities: {},
        clientInfo: {
          name: 'docent-web-ui',
          version: '0.1.0',
        },
      },
    };

    const { sessionId, data } = await this._post(body);

    this.sessionId = sessionId;
    this.protocolVersion = data.result.protocolVersion;

    await this._notifyInitialized();

    this.connected = true;
    return data.result;
  }

  async _notifyInitialized() {
    const body = {
      jsonrpc: '2.0',
      method: 'notifications/initialized',
    };

    await this._post(body);
  }

  async listTools() {
    const body = {
      jsonrpc: '2.0',
      id: ++this.requestId,
      method: 'tools/list',
    };
    const { data } = await this._post(body);
    return data;
  }

  async callTool(name, args) {
    const body = {
      jsonrpc: '2.0',
      id: ++this.requestId,
      method: 'tools/call',
      params: { name, arguments: args },
    };

    const { data } = await this._post(body);
    return data;
  }

  async _post(body) {
    const headers = {
      'Content-Type': 'application/json',
      'Accept': 'application/json, text/event-stream',
    };

    if (this.sessionId) {
      headers['Mcp-Session-Id'] = this.sessionId;
    }
    if (this.protocolVersion) {
      headers['MCP-Protocol-Version'] = this.protocolVersion;
    }

    const response = await fetch('/', {
      method: 'POST',
      headers,
      body: JSON.stringify(body),
    });

    const sessionId = response.headers.get('Mcp-Session-Id');

    if (response.status === 202) {
      return { sessionId, data: null };
    }

    if (!response.ok) {
      const text = await response.text().catch(() => '');
      throw new Error(`HTTP ${response.status}: ${text || response.statusText}`);
    }

    const contentType = response.headers.get('Content-Type') || '';

    let data;
    if (contentType.includes('text/event-stream')) {
      data = await this._readSseEvent(response);
    } else {
      data = await response.json();
    }

    return { sessionId, data };
  }

  async _readSseEvent(response) {
    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = '';

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });

      const result = extractFirstDataEvent(buffer);
      if (result.event) {
        reader.cancel();
        return result.event;
      }
      buffer = result.remainder;
    }

    throw new Error('No response event in SSE stream');
  }
}

File: e2e-tests/mcp_client.py
"""MCP streamable HTTP client helpers for e2e tests."""

from __future__ import annotations

import json
from typing import Any

import requests


PROTOCOL_VERSION = "2025-11-25"


def _extract_json(text: str) -> dict[str, Any]:
    if "data:" in text:
        data_lines = [
            line.removeprefix("data:").strip()
            for line in text.splitlines()
            if line.startswith("data:")
        ]
        if not data_lines:
            raise RuntimeError(f"No data: lines in SSE response: {text}")
        return json.loads(data_lines[-1])
    return json.loads(text)


def send_mcp_request(
    base_url: str,
    client: requests.Session,
    method: str,
    params: dict[str, Any] | None = None,
    session_id: str | None = None,
) -> dict[str, Any]:
    body: dict[str, Any] = {
        "jsonrpc": "2.0",
        "id": 1,
        "method": method,
    }
    if params is not None:
        body["params"] = params

    headers: dict[str, str] = {
        "Content-Type": "application/json",
        "Accept": "application/json, text/event-stream",
    }
    if session_id is not None:
        headers["mcp-session-id"] = session_id

    resp = client.post(base_url, json=body, headers=headers, timeout=30)
    resp.raise_for_status()
    return _extract_json(resp.text)


def initialize(base_url: str, client: requests.Session) -> tuple[dict[str, Any], str]:
    body = {
        "jsonrpc": "2.0",
        "id": 1,
        "method": "initialize",
        "params": {
            "protocolVersion": PROTOCOL_VERSION,
            "capabilities": {},
            "clientInfo": {"name": "e2e-test", "version": "0.1.0"},
        },
    }
    headers = {
        "Content-Type": "application/json",
        "Accept": "application/json, text/event-stream",
    }
    resp = client.post(base_url, json=body, headers=headers, timeout=30)
    resp.raise_for_status()

    session_id = resp.headers.get("mcp-session-id")
    assert session_id is not None, "initialize response missing mcp-session-id"

    return _extract_json(resp.text), session_id


def search(
    base_url: str,
    client: requests.Session,
    session_id: str,
    query: str,
    search_path: str = "/**",
) -> list[dict[str, Any]]:
    resp = send_mcp_request(
        base_url,
        client,
        "tools/call",
        {
            "name": "search_doc",
            "arguments": {
                "query": query,
                "limit": 5,
                "search_path": search_path,
            },
        },
        session_id=session_id,
    )
    text = resp["result"]["content"][0]["text"]
    return json.loads(text)
